En esta jupyter notebook se desarrolla la extracción, transformación y carga del conjunto de datos para 2019.

In [1]:
import pandas as pd
import json

%load_ext autoreload
%autoreload 2

import utils

import warnings
warnings.filterwarnings("ignore")

Se extraen los datos desde el archivo .csv, se convierte en Dataframe y se observa su contenido.

In [2]:
# Ruta al dataset
ruta_siniestro2020 = '../Datasets/feu-siniestros - (FC)feu-siniestros-2020.csv'

# Leer el archivo CSV directamente en un DataFrame
df_sin2020 = pd.read_csv(ruta_siniestro2020)

# Mostrar el DataFrame
df_sin2020

id_feu    unidad_regional siniestro_fecha siniestro_hora   provincia  \
0     71290           OBS VIAL      2020-12-28       17:34:00  CORRIENTES   
1     75820  OBSERVATORIO VIAL      2020-12-27       12:15:00  CORRIENTES   
2     72215          OBS. VIAL      2020-12-27       19:22:00  CORRIENTES   
3     79464  OBSERVATORIO VIAL      2020-12-26        7:14:00  CORRIENTES   
4     84248  OBSERVATORIO VIAL      2020-12-25       19:25:00  CORRIENTES   
..      ...                ...             ...            ...         ...   
433   74033  OBSERVATORIO VIAL      2020-01-07       12:15:00  CORRIENTES   
434   87513  OBSERVATORIO VIAL      2020-01-06        8:33:00  CORRIENTES   
435   81359  OBSERVATORIO VIAL      2020-01-04       20:50:00  CORRIENTES   
436   90039  OBSERVATORIO VIAL      2020-01-03        7:00:00  CORRIENTES   
437   73999  OBSERVATORIO VIAL      2020-01-03       13:35:00  CORRIENTES   

     id_provincia departamento  id_departamento   localidad zona_ocurrencia  \
0              18      CAPITAL               21  CORRIENTES          Urbana   
1              18      CAPITAL               21  CORRIENTES          Urbana   
2              18      CAPITAL               21  CORRIENTES           Rural   
3              18      CAPITAL               21  CORRIENTES          Urbana   
4              18      CAPITAL               21  CORRIENTES          Urbana   
..            ...          ...              ...         ...             ...   
433            18      CAPITAL               21  CORRIENTES          Urbana   
434            18      CAPITAL               21  CORRIENTES          Urbana   
435            18      CAPITAL               21  CORRIENTES          Urbana   
436            18      CAPITAL               21  CORRIENTES          Urbana   
437            18      CAPITAL               21  CORRIENTES          Urbana   

     ... senalizacion_carteleria senalizacion_horizontal  \
0    ...                     NaN                     NaN   
1    ...                     S/D                     S/D   
2    ...                     NaN                     NaN   
3    ...                      Si                      No   
4    ...                      Si                      No   
..   ...                     ...                     ...   
433  ...                     S/D                     S/D   
434  ...                      Si                      Si   
435  ...                     S/D                     S/D   
436  ...                      Si                      Si   
437  ...                      Si                     S/D   

    senalizacion_transitoria luminosidad luz_artificial transito_restringido  \
0                        NaN         Día            NaN                  NaN   
1                        S/D         Día             NO      Sin Restricción   
2                        NaN         Día             SI                  NaN   
3                         No         Día             NO                  NaN   
4                         No         Día             NO      Sin Restricción   
..                       ...         ...            ...                  ...   
433                      S/D         Día             NO      Sin Restricción   
434                      NaN         Día             NO      Sin Restricción   
435                      S/D       Noche             SI      Sin Restricción   
436                       No         Día             NO      Sin Restricción   
437                      S/D         Día             NO      Sin Restricción   

    estado_ambiental  visibilidad  visibilidad_otro      semaforo  
0          Despejado          NaN               NaN           NaN  
1          Despejado          NaN               NaN  Sin Semáforo  
2          Despejado        Buena               NaN           NaN  
3          Despejado        Buena               NaN      Funciona  
4          Despejado        Buena               NaN  Sin Semáforo  
..               ...          ...        

Se revisan los tipos de datos por columna y la cantidad de nulos.

In [3]:
utils.verificar_tipo_datos(df_sin2020)

nombre_campo                        tipo_datos  \
0                              id_feu                   [<class 'int'>]   
1                     unidad_regional                   [<class 'str'>]   
2                     siniestro_fecha                   [<class 'str'>]   
3                      siniestro_hora                   [<class 'str'>]   
4                           provincia                   [<class 'str'>]   
5                        id_provincia                   [<class 'int'>]   
6                        departamento                   [<class 'str'>]   
7                     id_departamento                   [<class 'int'>]   
8                           localidad  [<class 'str'>, <class 'float'>]   
9                     zona_ocurrencia  [<class 'str'>, <class 'float'>]   
10                        via_publica                   [<class 'str'>]   
11                         nombre_via                   [<class 'str'>]   
12                          altura_km  [<class 'str'>, <class 'float'>]   
13                      entre_calle_1  [<class 'str'>, <class 'float'>]   
14                      entre_calle_2  [<class 'float'>, <class 'str'>]   
15                            latitud  [<class 'str'>, <class 'float'>]   
16                           longitud  [<class 'str'>, <class 'float'>]   
17                             ilesos                   [<class 'int'>]   
18                            heridos                   [<class 'int'>]   
19                         fallecidos                   [<class 'int'>]   
20                          vehiculos                   [<class 'int'>]   
21                           peatones                   [<class 'int'>]   
22            tipo_siniestro_multiple  [<class 'str'>, <class 'float'>]   
23               tipo_siniestro_unico  [<class 'str'>, <class 'float'>]   
24                    despiste_previo                   [<class 'str'>]   
25                tipo_siniestro_otro  [<class 'float'>, <class 'str'>]   
26                      tipo_colision  [<class 'str'>, <class 'float'>]   
27           cantidad_de_involucrados                   [<class 'int'>]   
28                        trazado_via                   [<class 'str'>]   
29            configuracion_de_la_via  [<class 'float'>, <class 'str'>]   
30  configuracion_de_la_via_otro_tipo  [<class 'float'>, <class 'str'>]   
31                  pendiente_declive  [<class 'float'>, <class 'str'>]   
32                     cono_de_vision                 [<class 'float'>]   
33             material_de_la_calzada  [<class 'str'>, <class 'float'>]   
34        material_de_la_calzada_otro                 [<class 'float'>]   
35               estado_de_la_calzada  [<class 'str'>, <class 'float'>]   
36            estado_fisico_ambiental  [<class 'str'>, <class 'float'>]   
37        estado_de_conservacion_otro  [<class 'float'>, <class 'str'>]   
38          division_fisica_de_la_via  [<class 'float'>, <class 'str'>]   
39     division_fisica_de_la_via_otro  [<class 'float'>, <class 'str'>]   
40              barreras_de_seguridad                 [<class 'float'>]   
41                          banquinas                 [<class 'float'>]   
42            senalizacion_carteleria  [<class 'float'>, <class 'str'>]   
43            senalizacion_horizontal  [<class 'float'>, <class 'str'>]   
44           senalizacion_transitoria  [<class 'float'>, <class 'str'>]   
45                        luminosidad  [<class 'str'>, <class 'float'>]   
46                     luz_artificial  [<class 'float'>, <class 'str'>]   
47               transito_restringido  [<class 'float'>, <class 'str'>]   
48                   estado_ambiental  [<class 'str'>, <class 'float'>]   
49                        visibilidad  [<class 'float'>, <class 'str'>]   
50                   visibilidad_otro  [<class 'float'>, <class 'str'>]   
51                           semaforo  [<class 'float'>, <class 'str'>]   

    no_nulos_%  nulos_%  nulos  
0       100.00  

Se observa que hay columnas con más de 90% de nulos, por lo cual decidimos borrarlas.

In [4]:
# Eliminar columnas con más del 80% de valores nulos
umbral = 0.8  # Umbral del 80%
df_2020filtrado = df_sin2020.dropna(thresh=len(df_sin2020) * (1 - umbral), axis=1)

# Mostrar las columnas que se mantuvieron
df_2020filtrado.columns

Index(['id_feu', 'unidad_regional', 'siniestro_fecha', 'siniestro_hora',
       'provincia', 'id_provincia', 'departamento', 'id_departamento',
       'localidad', 'zona_ocurrencia', 'via_publica', 'nombre_via',
       'altura_km', 'entre_calle_1', 'latitud', 'longitud', 'ilesos',
       'heridos', 'fallecidos', 'vehiculos', 'peatones',
       'tipo_siniestro_multiple', 'tipo_siniestro_unico', 'despiste_previo',
       'tipo_colision', 'cantidad_de_involucrados', 'trazado_via',
       'configuracion_de_la_via', 'pendiente_declive',
       'material_de_la_calzada', 'estado_de_la_calzada',
       'estado_fisico_ambiental', 'division_fisica_de_la_via',
       'senalizacion_carteleria', 'senalizacion_horizontal',
       'senalizacion_transitoria', 'luminosidad', 'luz_artificial',
       'transito_restringido', 'estado_ambiental', 'visibilidad', 'semaforo'],
      dtype='object')

Verificamos ahora duplicados

In [5]:
utils.verifica_duplicados_por_columna(df_2020filtrado, 'id_feu')

'No hay duplicados'

No hay duplicados, podemos continuar corrigiendo los tipos de datos en algunas columnas. Por ejemplo, fechas y horas.

In [6]:
# Convertir la columna siniestro_fecha a tipo datetime
df_2020filtrado['siniestro_fecha'] = pd.to_datetime(df_2020filtrado['siniestro_fecha'], format='%Y-%m-%d', errors='coerce')

# Convertir la columna siniestro_hora a tipo time (si está en formato 24 horas)
df_2020filtrado['siniestro_hora'] = pd.to_datetime(df_2020filtrado['siniestro_hora'], format='%H:%M:%S', errors='coerce').dt.time

# Verificar las conversiones
df_2020filtrado[['siniestro_fecha', 'siniestro_hora']].head()

siniestro_fecha siniestro_hora
0      2020-12-28       17:34:00
1      2020-12-27       12:15:00
2      2020-12-27       19:22:00
3      2020-12-26       07:14:00
4      2020-12-25       19:25:00

Ahora, podemos pasar en limpio qué datos contiene este Dataframe

In [7]:
utils.verificar_tipo_datos(df_2020filtrado)

nombre_campo  \
0                      id_feu   
1             unidad_regional   
2             siniestro_fecha   
3              siniestro_hora   
4                   provincia   
5                id_provincia   
6                departamento   
7             id_departamento   
8                   localidad   
9             zona_ocurrencia   
10                via_publica   
11                 nombre_via   
12                  altura_km   
13              entre_calle_1   
14                    latitud   
15                   longitud   
16                     ilesos   
17                    heridos   
18                 fallecidos   
19                  vehiculos   
20                   peatones   
21    tipo_siniestro_multiple   
22       tipo_siniestro_unico   
23            despiste_previo   
24              tipo_colision   
25   cantidad_de_involucrados   
26                trazado_via   
27    configuracion_de_la_via   
28          pendiente_declive   
29     material_de_la_calzada   
30       estado_de_la_calzada   
31    estado_fisico_ambiental   
32  division_fisica_de_la_via   
33    senalizacion_carteleria   
34    senalizacion_horizontal   
35   senalizacion_transitoria   
36                luminosidad   
37             luz_artificial   
38       transito_restringido   
39           estado_ambiental   
40                visibilidad   
41                   semaforo   

                                           tipo_datos  no_nulos_%  nulos_%  \
0                                     [<class 'int'>]      100.00     0.00   
1                                     [<class 'str'>]      100.00     0.00   
2   [<class 'pandas._libs.tslibs.timestamps.Timest...      100.00     0.00   
3                           [<class 'datetime.time'>]      100.00     0.00   
4                                     [<class 'str'>]      100.00     0.00   
5                                     [<class 'int'>]      100.00     0.00   
6                                     [<class 'str'>]      100.00     0.00   
7                                     [<class 'int'>]      100.00     0.00   
8                    [<class 'str'>, <class 'float'>]       99.77     0.23   
9                    [<class 'str'>, <class 'float'>]       99.77     0.23   
10                                    [<class 'str'>]      100.00     0.00   
11                                    [<class 'str'>]      100.00     0.00   
12                   [<class 'str'>, <class 'float'>]       99.77     0.23   
13                   [<class 'str'>, <class 'float'>]       80.59    19.41   
14                   [<class 'str'>, <class 'float'>]       94.98     5.02   
15                   [<class 'str'>, <class 'float'>]       94.98     5.02   
16                                    [<class 'int'>]      100.00     0.00   
17                                    [<class 'int'>]      100.00     0.00   
18                                    [<class 'int'>]      100.00     0.00   
19                                    [<class 'int'>]      100.00     0.00   
20                                    [<class 'int'>]      100.00     0.00   
21                   [<class 'str'>, <class 'float'>]       99.09     0.91   
22                   [<class 'str'>, <class 'float'>]       99.09     0.91   
23                                    [<class 'str'>]      100.00     0.00   
24                   [<class 'str'>, <class 'float'>]       39.73    60.27   
25                                    [<class 'int'>]      100.00     0.00   
26                                    [<class 'str'>]      100.00     0.00   
27                   [<class 'float'>, <class 'str'>]       89.50    10.50   
28                   [<class 'float'>, <class 'str'>]       95.89     4.11   
29                   [<class 'str'>, <class 'float'>]       97.03     2.97   
30                   [<class 'str'>, <class 'float'>]       96.35     3.65   
31                   [<class 'str'>, <class 'float'>]       97.26     2.74   
32                   [<

In [8]:
print(df_2020filtrado.shape)

(438, 42)


Este conjunto contiene 42 columnas y quedó con 438 filas luego de borrar las columnas con mayor porcentaje de nulos. Las columnas que contiene son:

* id_feu: es el índice del siniestro.
* unidad_regional: es el nombre de la unidad regional encargada de cargar el dato. En este caso, el 100% corresponde a Observatorio Vial.
* siniestro_fecha: es la fecha del siniestro.
* siniestro_hora: es la hora del siniestro.
* id_provincia: es la provincia en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* departamento: es el departamento en el que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Capital.
* id_departamento: es la ID del departamento en el que ocurrió el siniestro.
* localidad: es la localidad en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* zona_ocurrencia: es la especificación sobre el tipo de zona donde ocurrió el siniestro. Urbana o Rural.
* via_publica: es la especificación sobre el tipo de calle en la que ocurrió el siniestro. Avenida, Calle o Ruta.
* nombre_via: es el nombre de la calle, avenida o ruta donde ocurrió el siniestro.
* altura_km: es la altura en kilómetros dónde ocurrió el siniestro.
* entre_calle_1: es el cruce en donde ocurrió el siniestro.
* latitud: es la latitud de dónde ocurrió el siniestro.
* longitud: es la longitud de dónde ocurrió el siniestro.
* ilesos: es la cantidad de personas que no sufrieron daños en el siniestro.
* heridos: es la cantidad de personas que sufrieron daños en el siniestro.
* fallecidos: es la cantidad de personas que fallecieron en el siniestro.
* vehiculos: es la cantidad de vehículos involucrados en el siniestro.
* peatones: es la cantidad de peatones involucrados en el siniestro.
* tipo_siniestro_multiple: es la especificación del hecho.
* despiste_previo: es la respuesta ante el hecho.
* cantidad_involucrados: es la cantidad de personas que participaron del siniestro.
* trazado_via: es la especficidad sobre el tipo de calle en que ocurrió el hecho. Recta o Curva.
* configuracion_de_la_via: es el tipo de configuración de la vía. Intersección, Cruce, Rotonda.
* pendiente_declive: es si existe pendiente.
* material_de_la_calzada: es el material del que está hecha la calzada donde ocurrió el siniestro.
* estado_de_la_calzada: es el estado en que se encontraba la calzada en el momento del siniestro.
* estado_fisico_ambiental: el estado de la calzada al momento del siniestro. Seca, Mojada.
* division_fisica_de_la_via: es el qué divide la vía donde ocurrió el siniestro. Por ejemplo, barrera de hormigón.
* senalización_carteleria: es si existía señalización en la zona al momento del hecho.
* senalización_horizontal: es si existía señalización horizontal en la zona al momento del hecho.
* senalización_transitoria: es si existía señalización transitoria en la zona al momento del hecho.
* luminosidad: es si el hecho ocurrió de día o de noche. Luz Natural.
* luz_artifical: es si en el lugar del hecho existía luz artificial.
* transito_restringido: es si el lugar era de tránsito restringido.
* estado_ambiental: es cómo estaba el tiempo al momento del hecho. Despejado, lluvioso, etc.
* visibilidad: es cómo se encontraba la visibilidad al momento del hecho. 
* semaforo: es si existe semáforo en la zona del hecho, y su funcionamiento.

In [9]:
# Eliminar columnas innecesarias
columnas_a_eliminar = [
    'unidad_regional',
    'id_provincia',
    'id_departamento',
    'altura_km',
    'tipo_siniestro_multiple',
    'despiste_previo',
    'tipo_siniestro_unico',
    'trazado_via',
    'configuracion_de_la_via',
    'pendiente_declive',
    'division_fisica_de_la_via',
    'senalizacion_carteleria',
    'senalizacion_horizontal',
    'senalizacion_transitoria',
    'transito_restringido', 
    'provincia',
    'departamento',
    'localidad',
    ]

df_2020filtrado = df_2020filtrado.drop(columns=columnas_a_eliminar)


Ahora tenemos un Dataframe mucho más manejable, con los datos más importantes para nuestra investigación. Vamos a guardarlo para su futuro uso.

In [10]:
sin2020 = '../Datasets_limpios/siniestros2020.csv'
df_2020filtrado.to_csv(sin2020)